Semester Refactor

liuric opened the pull request:
Refactored semester system to rely on a class, and provided options for adding/removing more years.
Currently adding years isn't permanent/saved onto road unless the user adds a course in that year, and the minimum number of years is 4.

liuric committed:
Initial Semester Refactor Commit

liuric committed:
Fixed crash on open/adding course, no items appearing in recyclerview, and semester year upper limit not updating properly

liuric committed:
Refactored Semester to rely on RoadDocument for the # of years in the current road
Added buttons to myRoadFragment to add/remove years
Fixed Prior Credit Button having opposite behavior as desired (enabled when course already in)

liuric committed:
Fixed year modifying cell not taking up entire width
Allowed addCourseDialog to update recyclerview based on years

liuric committed:
Modified oldSemesterIndex to return a valid index
(or stuff it into prior credit, if there exists no valid
index for the new semester system)

venkatesh-sivaraman commented:
Testing out the app (haven't looked at the code yet), it looks great! I just found two bugs: 

1. When opening a blank road document and adding a course (6.003 to 2nd year fall), I get the following crash:

```
java.lang.NullPointerException: Attempt to invoke interface method 'java.util.Map com.base12innovations.android.fireroad.models.doc.Semester$Delegate.getSemesterNames()' on a null object reference
        at com.base12innovations.android.fireroad.models.doc.Semester.toString(Semester.java:245)
        at com.base12innovations.android.fireroad.adapter.MyRoadCoursesAdapter.formatSectionHeader(MyRoadCoursesAdapter.java:197)
        at com.base12innovations.android.fireroad.adapter.CourseCollectionAdapter.onBindViewHolder(CourseCollectionAdapter.java:97)
        at com.base12innovations.android.fireroad.adapter.MyRoadCoursesAdapter.onBindViewHolder(MyRoadCoursesAdapter.java:316)
        at com.base12innovations.android.fireroad.adapter.MyRoadCoursesAdapter.onBindViewHolder(MyRoadCoursesAdapter.java:27)
        at android.support.v7.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:6735)
        at android.support.v7.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:6777)
        at android.support.v7.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:5706)
        at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:5973)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5812)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5808)
        at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2230)
        at android.support.v7.widget.GridLayoutManager.layoutChunk(GridLayoutManager.java:557)
        at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1517)
        at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:587)
        at android.support.v7.widget.GridLayoutManager.onLayoutChildren(GridLayoutManager.java:171)
        at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3878)
        at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3595)
        at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:4148)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.support.constraint.ConstraintLayout.onLayout(ConstraintLayout.java:1858)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.support.constraint.ConstraintLayout.onLayout(ConstraintLayout.java:1858)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.support.design.widget.CoordinatorLayout.layoutChild(CoordinatorLayout.java:1183)
        at android.support.design.widget.CoordinatorLayout.onLayoutChild(CoordinatorLayout.java:870)
        at android.support.design.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:889)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.support.v4.widget.DrawerLayout.onLayout(DrawerLayout.java:1233)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
2020-04-04 12:07:09.680 6612-6612/? E/AndroidRuntime:     at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
        at com.android.internal.policy.DecorView.onLayout(DecorView.java:779)
        at android.view.View.layout(View.java:21912)
        at android.view.ViewGroup.layout(ViewGroup.java:6260)
        at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3080)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2590)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1721)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7598)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:966)
        at android.view.Choreographer.doCallbacks(Choreographer.java:790)
        at android.view.Choreographer.doFrame(Choreographer.java:725)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:951)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

2. When I create a new road from the document browser and try to add a course, the labels in the Add Course dialog are cut off and you can't select a semester to add to. If I open a different document that I created previously, however, I can see the dialog properly.

Will follow up with more comments later :)

liuric committed:
Fix for crash upon creating new RoadDocument

venkatesh-sivaraman commented on app/src/main/java/com/base12innovations/android/fireroad/dialog/AddCourseDialog.java:
```diff
@@ -124,4 +113,19 @@ public void onClick(DialogInterface dialogInterface, int i) {
+    public boolean courseInSemester(Semester semester){
+        return doc != null && doc.coursesForSemester(semester).contains(course);
+    }
+    public boolean courseNotOfferedInSemester(Semester semester){
```
Can this be a method on `Course`? It could be useful in other contexts

liuric replied:
Done.

venkatesh-sivaraman commented on app/src/main/java/com/base12innovations/android/fireroad/models/doc/RoadDocument.java:
```diff
@@ -614,4 +621,64 @@ public void removeProgressOverride(String keyPath){
     }
+
+    private int numYears;
+    private Map<Semester, String> semesterNames = new LinkedHashMap<>();
```
I like this use of `LinkedHashMap`! However, I'm not sure why we have to store semesters as maps from semesters to names and IDs - shouldn't name and ID be properties of the `Semester` object itself? In that case we could just store an `ArrayList` of semester objects and grab their names and IDs whenever we want.

If you do want to keep using linked hash map, could you add a short docstring explaining that correct behavior *depends* on the use of `LinkedHashMap` (i.e. iteration order equals insertion order)? And maybe make these maps `final` in case some unknowing helper later tries to reassign it to a different kind of `Map`?

liuric replied:
I removed the LInkedHashMaps and just used a ArrayList for ordered iteration through semesters

venkatesh-sivaraman commented on app/src/main/java/com/base12innovations/android/fireroad/models/doc/Semester.java:
```diff
@@ -0,0 +1,272 @@
+        init(year,season,false);
+    }
+    private void init(int year, Season season, boolean updateNumYears){
+        init(year,season,updateNumYears,new WeakReference<Delegate>(User.currentUser().getCurrentDocument()));
```
Let's avoid having the `Semester` class know how to access the global current document

liuric replied:
Done.

venkatesh-sivaraman commented on app/src/main/java/com/base12innovations/android/fireroad/models/doc/Semester.java:
```diff
@@ -0,0 +1,272 @@
+        this.delegate = delegate;
+        if(updateNumYears)
+            delegate.get().updateNumYears(year);
+        isValid = year > 0 && year <= delegate.get().getNumYears();
```
Why have the `Semester` object exist at all if it's not valid? Since the road document is the one that determines whether a semester is valid, can that be a method on the road document (i.e. `RoadDocument.isSemesterValid(Semester semester)` instead?

liuric replied:
The nextSemester() and prevSemester() methods, as well as creation of the object directly from the semesterID can generate "invalid" semesters.
I moved this to the road document as suggested.

venkatesh-sivaraman commented on app/src/main/java/com/base12innovations/android/fireroad/models/doc/Semester.java:
```diff
@@ -0,0 +1,272 @@
+
+    @Override
+    public String toString(){
+        if(delegate != null && delegate.get() != null && delegate.get().getSemesterNames().containsKey(this))
```
Why does the semester have to get its name from the delegate? It seems like the semester should own that information, and the road document shouldn't even have a map for semester names. Rather, it should just call `toString` on the semester object every time it needs to display the semester name.

liuric replied:
Yup, removed.

venkatesh-sivaraman reviewed (changes requested):
A couple more UI nitpicks, and then I've made a few suggestions on the code structure :)

1. There should be more padding below the "Add New Year" button, since it is often covered up by the collapsed course detail panel. Also, this button should probably be renamed to "Add Another Year" since we're not actually creating time :P
2. The buttons in the `SelectSemesterAdapter` are currently not visually centered because the words have different lengths - I guess this wasn't as obvious when there were only three buttons, but now it is rather visible. So instead of making the buttons equal layout weight in the `LinearLayout`, you can add equal-weight spacers between the buttons (see [this stackoverflow](https://stackoverflow.com/a/14125793)).
3. This is definitely optional, but if you have time it would be nice to make the "Remove Last Year" always clickable, but display an error alert if there are subjects in the last year. It can be confusing when buttons are disabled for no apparent reason..

Sorry for the delay in this review - awesome work!

liuric committed:
Added spacing in between semesters in AddCourseDialog (still looks a little off though)
Added spacing below add/remove year buttons
Removed delegate, isValid fields from Semester
Replaced LinkedHashMap uses with ArrayList for purposes of iterating through semesters
Added inSemester method to Course
Added Dialog Warnings when attempting to remove years when there are only 4/last year still has courses and removed disabling of remove year button

venkatesh-sivaraman reviewed (approved):
This is super close to being ready! I've been implementing the iOS version following the strategies you used, and the main thing that came up was that we need to save `numYears` in the road file. Otherwise, you can add as many years as you want, but if you don't add courses to them, they won't get saved. I just declared a top-level JSON key in the road file called `numYears` - can you implement that in the road read/write methods? I think the default value should be 5 if the key isn't present, although we should initialize new roads with 4 years as you're doing already.

Just one other tiny nitpick: the "Remove Last Year" button is doing exactly what I was envisioning now, but it would be great if it could appear disabled when the action isn't allowed (by reducing the opacity). That way, the button visually hints to the user that the action will fail, but if the user still taps the button, it explains *why* the action failed.

After you've made these changes, please go ahead and merge!!

liuric committed:
Added numYears field to RoadDocument JSON, changed default years for previous roads to be 5, added greying out of remove Year button and saving when adding/removing years.
